src/number_theory/bernoulli.lean

```diff
@@ -139,8 +164,48 @@ begin
   rw ← cast_sub hk,
   congr',
   field_simp [show ((n - k : ℕ) : ℚ) + 1 ≠ 0, by {norm_cast, simp}],
-  -- down to nat
   norm_cast,
   rw [mul_comm, nat.sub_add_eq_add_sub hk],
   exact choose_mul_succ_eq n k,
 end
+
+open power_series
+open nat
+
+@[simp] lemma constant_coeff_exp : constant_coeff ℚ (exp ℚ) = 1 := rfl
+
+@[simp] theorem rat.rat_algebra_map_rat (q : ℚ) : algebra_map ℚ ℚ q = q :=
+begin
+  rw [show algebra_map ℚ ℚ = (by refine_struct { to_fun := id}; tidy),
+    by simp only [eq_iff_true_of_subsingleton]], refl,
+end
+
+theorem bernoulli_power_series :
+(power_series.mk (λ n, ((bernoulli n) / (nat.factorial n) : ℚ))) * (power_series.exp ℚ - 1) =
+  (X : power_series ℚ) * (exp ℚ) :=
+begin
+  ext n,
+  -- constant coefficient is a special case
+  cases n,
+  simp only [ring_hom.map_sub, constant_coeff_one, zero_mul, constant_coeff_exp, constant_coeff_X,
+  coeff_zero_eq_constant_coeff, mul_zero, sub_self, ring_hom.map_mul],
+  rw [coeff_mul, mul_comm X, coeff_succ_mul_X],
+  simp only [coeff_mk, coeff_one, coeff_exp, linear_map.map_sub, factorial,
+  rat.rat_algebra_map_rat], rw nat.sum_antidiagonal_succ',
+  simp, --squeeze_simp hangs
+  apply eq_inv_of_mul_left_eq_one,
+  rw sum_mul,
+  convert bernoulli_spec' n using 1,
+  apply sum_congr rfl,
+  rintro ⟨i, j⟩ hn, rw nat.mem_antidiagonal at hn, subst hn, dsimp only,
+  have hj : (j : ℚ) + 1 ≠ 0, by norm_cast; linarith,
```

Comment:
Here a comma instead of a semicolon suffices, isn't it?

Reply:
Yes, please write `by { norm_cast, linarith },`

Reply:
Done, thank you!